src/components/verification-grid/verification-grid.ts

```diff
@@ -292,6 +338,194 @@ export class VerificationGridComponent extends AbstractComponent(LitElement) {
+
+      // sometimes we can get stuck in an infinite loop where increasing grid
+      // size will result in the grid not fitting, but decreasing the grid size
+      // will result in a grid that does not meet the aspect ratio threshold
+      // to resolve this, we keep a record of the targets that we have checked
+      // if we see the same one twice (we have increased, then decreased)
+      // we will decrement the grid size until we find a grid that fits
+      if (testedTargets.includes(refinedTarget)) {
+        refinedTarget -= 1;
+        continue;
+      }
+      testedTargets.push(refinedTarget);
+
+      let hasFittingCandidate = false;
+      for (let i = 1; i <= refinedTarget; i++) {
+        if (refinedTarget % i === 0) {
+          const columns = i;
+          const rows = refinedTarget / i;
+
+          const candidateShape = { columns, rows };
+          const willFit = this.willFitTileSize(containerSize, candidateShape);
+          const meetsAspectRatio = this.gridAspectRatioSimilarity(containerSize, candidateShape) > targetThreshold;
+
+          if (willFit) {
+            hasFittingCandidate = true;
+
+            if (meetsAspectRatio) {
+              candidateShapes.push({ columns, rows });
+            }
+          }
+        }
+      }
+
+      // if we have reached this condition, there is a fatal bug in the
+      // dynamic grid size code. This condition should never be triggered
+      // I have added the condition so that if dynamic grid size gets stuck in
+      // and infinite loop it won't completely crash the browser
+      // meaning that the verification grid will soft fail to prevent all the
+      // users work being lost
+      const maxGridSize = 1_000 as const;
+      if (refinedTarget >= maxGridSize || refinedTarget <= 0) {
+        console.error("Reached maximum/minimum grid size. Recovering via a 1x1 grid.");
+        this.targetGridSize = 1;
+        return { columns: 1, rows: 1 };
+      }
+
+      // there are two reasons why we might reject candidate shapes
+      // either they would not fit on the screen or they do not meet the
+      // minimum similarity to the containers aspect ratio
+      //
+      // if all the candidates are rejected because they do not fit, we will
+      // attempt to decrease the size until we find a target that correctly fits
+      //
+      // if we had a candidate that fit, but we rejected it because it did not
+      // meet the aspect ratio threshold, we continue to increase the target
+      // until we find a candidate that meets our conditions
+      if (candidateShapes.length === 0) {
+        if (!hasFittingCandidate) {
+          if (isPrime(refinedTarget)) {
+            refinedTarget += 1;
+            continue;
+          }
+
+          if (refinedTarget === 1) {
+            return { columns: 1, rows: 1 };
+          }
+
+          refinedTarget -= 1;
+          continue;
+        }
+
+        refinedTarget += 1;
+        continue;
+      }
```

Comment:
i recognize this in unhelpful but this gives me the heeby jeebies. See above comment about recursive functions